client-spark/common/src/main/java/org/apache/spark/shuffle/celeborn/SortBasedPusher.java

```diff
@@ -44,6 +44,64 @@
 
 public class SortBasedPusher extends MemoryConsumer {
 
+  class MemoryThresholdManager {
+
+    private final long maxMemoryThresholdInBytes;
+    private final double smallPushTolerateFactor;
+
+    private final long sendBufferSizeInBytes;
+
+    MemoryThresholdManager(
+        int numPartitions, long sendBufferSizeInBytes, double smallPushTolerateFactor) {
+      this.maxMemoryThresholdInBytes = numPartitions * sendBufferSizeInBytes;
+      this.smallPushTolerateFactor = smallPushTolerateFactor;
+      this.sendBufferSizeInBytes = sendBufferSizeInBytes;
+    }
+
+    private boolean shouldGrow() {
+      boolean enoughSpace = pushSortMemoryThreshold * 2 <= maxMemoryThresholdInBytes;
+      double expectedPushSize = Long.MAX_VALUE;
+      if (this.expectedPushedCount != 0) {
+        expectedPushSize = this.expectedPushedBytes * 1.0 / this.expectedPushedCount;
+      }
+      boolean tooManyPushed =
+          pushedMemorySizeInBytes * 1.0 / pushedCount * (1 + this.smallPushTolerateFactor)
+              < expectedPushSize;
+      return enoughSpace && tooManyPushed;
+    }
+
+    public void growThresholdIfNeeded() {
+      if (shouldGrow()) {
+        long oldThreshold = pushSortMemoryThreshold;
+        pushSortMemoryThreshold = pushSortMemoryThreshold * 2;
```

Comment:
pushSortMemoryThreshold = max(pushSortMemoryThreshold * 2, maxMemoryThresholdInBytes)

Reply:
I think you meant min()?

Reply:
Yeah, min